hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/TxnReservedCapacityCounterImpl.java

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.mapstore.writebehind;
+
+import com.hazelcast.map.ReachedMaxSizeException;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.internal.util.MapUtil.isNullOrEmpty;
+
+/**
+ * This class represents capacity counters for write behind
+ * queueing. Basically, we have 2 capacity counters. One
+ * is node-wide capacity counter, other one is record-store's
+ * reserved capacity counter. Depending on the context
+ * we operate on a single counter or both.
+ *
+ * Note that record-store's reserved capacity counter
+ * is only used with transactions and when {@link
+ * com.hazelcast.config.MapStoreConfig#writeCoalescing} is off.
+ *
+ * Every record-store has its own counter instance.
+ *
+ * @see TxnReservedCapacityCounter
+ */
+public class TxnReservedCapacityCounterImpl implements TxnReservedCapacityCounter {
+
+    private final ConcurrentMap<UUID, Long> reservedCapacityCountByTxId;
+    private final NodeWideUsedCapacityCounter nodeWideUsedCapacityCounter;
+
+    public TxnReservedCapacityCounterImpl(NodeWideUsedCapacityCounter nodeWideUsedCapacityCounter) {
+        this.nodeWideUsedCapacityCounter = nodeWideUsedCapacityCounter;
+        this.reservedCapacityCountByTxId = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Increments 2 counters for supplied txnId.
+     * One is record-store's reserved capacity counter and other
+     * one is node-wide capacity counter. When node-wide
+     * capacity counter is exceeded preconfigured node-wide
+     * limit this method throws {@link ReachedMaxSizeException}
+     *
+     * @throws ReachedMaxSizeException (only when backup is false)
+     */
+    @Override
+    public void increment(UUID txnId, boolean backup) {
+        reservedCapacityCountByTxId.compute(txnId, (ignored, currentCapacityCount) -> {
+            if (backup) {
+                nodeWideUsedCapacityCounter.add(1L);
+            } else {
+                nodeWideUsedCapacityCounter.addCapacityOrThrowException(1);
+            }
+            return currentCapacityCount == null ? 1L : (currentCapacityCount + 1L);
+        });
+    }
+
+    /**
+     * Increments 2 counters with reserved capacities per
+     * txnId. One is record-store's reserved capacity
+     * counter and other one is node-wide capacity counter.
+     *
+     * Note that this method is only used with migrations
+     * and it doesn't throw ReachedMaxSizeException.
+     *
+     * @param reservedCapacityPerTxnId reserved capacities per
+     *                                 txnId
+     */
+    @Override
+    public void copy(Map<UUID, Long> reservedCapacityPerTxnId) {
+        if (isNullOrEmpty(reservedCapacityPerTxnId)) {
+            return;
```

Comment:
Is it really a valid scenario or we silently ignore invalid state here?

Reply:
deleted `if` statement.